src/main/java/com/gw/tasks/TaskManager.java

```diff
@@ -206,46 +209,40 @@ public void arrive(Task t) {
    * @param history_id
    */
   public void stopTask(String history_id) {
-
     try {
-
+      // Stop tasks in waiting list
       synchronized (waitinglist) {
         Iterator<Task> iterator = waitinglist.iterator();
+        List<Task> tasksToRemove = new ArrayList<>();
 
         while (iterator.hasNext()) {
-
           GeoweaverProcessTask thet = (GeoweaverProcessTask) iterator.next();
-
           if (thet.getHistory_id().equals(history_id)) {
-
-            thet.endPrematurely();
-
-            waitinglist.remove(thet); // remove from waiting list
+                thet.endPrematurely(); // Immediately end the task
+                tasksToRemove.add(thet); // Mark for removal
+              }
+            }
+            waitinglist.removeAll(tasksToRemove); // Remove all tasks at once after iteration
```

Comment:
This is wrong. If you want to remove all at once, this line should be outside the while loop. This doesn't make sense.